lib/testing.dart

```diff
@@ -0,0 +1,206 @@
+    // fixed.
+    node ? "node" : Platform.executable;
+
+/// Arguments that can be passed to [Process.start] and similar APIs along with
+/// the executable returned by [executableRunner] to run [executable], which is
+/// the basename of an executable in `bin` (without ".dart").
+///
+/// If [node] is `true`, this and [executableRunner] will run a NodeJS process
+/// using the executable compiled by `pkg-npm-dev`. Otherwise, they'll run a
+/// Dart VM process using the executable compiled by `pkg-standalone-dev` if one
+/// exists and running from source otherwise.
+///
+/// This throws a [TestFailure] if it would run a compiled executable that's
+/// out-of-date relative to the pubspec or to source files in `lib/` or `bin/`.
+///
+/// For example:
+///
+/// ```dart
+/// import 'dart:io';
+///
+/// import 'package:cli_pkg/testing.dart' as pkg;
+/// import 'package:test/test.dart';
+///
+/// void main() {
+///   test("hello-world prints a message", () async {
+///     var result = await Process.run(
+///         pkg.executableRunner("hello-world"),
+///         pkg.executableArgs("hello-world"));
+///     expect(result.stdout, equals("Hello, world!\n");
+///   });
+/// }
+/// ```
+///
+/// Note that in practice it's usually easier to use [start].
+List<String> executableArgs(String executable, {bool node = false}) {
+  if (node) {
+    var js = p.absolute("build/npm/$executable.js");
+    _ensureExecutableUpToDate(executable, js, 'pkg-npm-dev');
+    return [js];
+  }
+
+  var snapshot = p.absolute("build/$executable.dart.snapshot");
+  if (File(snapshot).existsSync()) {
+    _ensureExecutableUpToDate(executable, snapshot, 'pkg-standalone-dev');
+    return [snapshot];
+  }
+
+  return [
+    "-Dversion=$version",
+    "--enable-asserts",
+    p.absolute("bin/$executable.dart")
+  ];
+}
+
+/// Ensures that the [executable]'s compiled output at [path] is up-to-date
+/// relative to the sources that generate it.
+void _ensureExecutableUpToDate(
+    String executable, String path, String commandToRun) {
+  if (_executableUpToDateCache.add(path)) {
+    ensureUpToDate(path, commandToRun, dependencies: ['bin/$executable.dart']);
+  }
+}
+
+/// Ensures that [path] (usually a compilation artifact) has been modified more
+/// recently than all this package's source files.
+///
+/// By default, this checks files in the `lib/` directory as well as
+/// `pubspec.lock`. Additional files or directories to check can be passed in
+/// [dependencies].
+///
+/// If [path] doesn't exist or is out of date, throws a [TestFailure]
+/// encouraging the user to run [commandToRun].
+void ensureUpToDate(String path, String commandToRun,
+    {Iterable<String> dependencies}) {
+  // Ensure path is relative so the error messages are more readable.
+  path = p.relative(path);
+  if (!File(path).existsSync()) {
+    throw TestFailure("$path does not exist. Run $commandToRun.");
+  }
+
+  var entriesToCheck = [
+    for (var dependency in [...?dependencies, "lib"])
```

Comment:
Style nit here and below: braces?

Reply:
Braces aren't allowed within collection literals, since these aren't really statements but rather modifications to the expressions afterwards.